In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import matplotlib.colors
from matplotlib.colors import LogNorm
import pandas
# import warnings
# warnings.simplefilter(action='ignore', category=FutureWarning)
# import cdms
import sys
# import scipy
import random
# import scipy.signal
import numpy as np
import pandas as pd
# from importlib import reload
sys.path.insert(0,'/home/jovyan/RQanalysis')
from RQ import RQReader, registercut
print(sys.path)

/cvmfs/cdms.opensciencegrid.org/releases/centos7/V04-08/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)


['/home/jovyan/RQanalysis', '/home/jovyan/benchmark_dataframes', '', '/cvmfs/cdms.opensciencegrid.org/releases/centos7/V04-08/lib/python3.7/site-packages', '/cvmfs/sft.cern.ch/lcg/releases/ROOT/v6.22.00-2773b/x86_64-centos7-gcc8-opt/lib', '/cvmfs/sft.cern.ch/lcg/views/LCG_98py3cu10/x86_64-centos7-gcc8-opt/python', '/cvmfs/sft.cern.ch/lcg/views/LCG_98py3cu10/x86_64-centos7-gcc8-opt/lib', '/cvmfs/sft.cern.ch/lcg/views/LCG_98py3cu10/x86_64-centos7-gcc8-opt/lib/python3.7/site-packages', '/cvmfs/sft.cern.ch/lcg/releases/Python/3.7.6-7a85e/x86_64-centos7-gcc8-opt/lib/python37.zip', '/cvmfs/sft.cern.ch/lcg/releases/Python/3.7.6-7a85e/x86_64-centos7-gcc8-opt/lib/python3.7', '/cvmfs/sft.cern.ch/lcg/releases/Python/3.7.6-7a85e/x86_64-centos7-gcc8-opt/lib/python3.7/lib-dynload', '/home/jovyan/.local/lib/python3.7/site-packages', '/cvmfs/sft.cern.ch/lcg/releases/Python/3.7.6-7a85e/x86_64-centos7-gcc8-opt/lib/python3.7/site-packages', '/cvmfs/sft.cern.ch/lcg/views/LCG_98py3cu10/x86_64-centos7-gcc8-

In [2]:
data_dir = '/cvmfs/data/CDMS/RQanalysis_testing/RQroot/'
series_numbers =['25220213_171932','25220214_092356','25220215_073923']
files = [f'{data_dir}OFResults_{i}.root' for i in series_numbers]

#Loading the data
data = RQReader(files,progressbar=True)
# data.describe(verbose =1)
data1 = data.getchunk(0)
data2 = data.getchunk(1)
data3 = data.getchunk(2)


Checking files: 100%|██████████| 3/3 [00:00<00:00,  4.45files/s]
/home/jovyan/RQanalysis/RQ/rqdata.py:357: FutureWarning: In a future version of pandas all arguments of DataFrame.dropna will be keyword-only
  df = df.replace(-999999, np.NaN).dropna(0, 'all').dropna(1, 'all')


In [3]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 100000 entries, (25220213171932, 200000, 1) to (25220213171932, 299999, 1)
Columns: 525 entries, Slope_NFC1_in to IntegralTail_NFC2_in
dtypes: float32(405), float64(18), int16(5), int32(97)
memory usage: 209.5 MB


In [4]:
data1.head()

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 0           1        9.468837e-08                   -0.024721   
               1           1       -2.323795e-06                   -0.023009   
               2           1        1.909242e-06                   -0.028883   
               3           1       -5.892106e-07                   -0.022784   
               4           1       -1.040180e-03                   -0.026172   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 0           1                   0.224138             0   
               1           1                   0.190206             0   
               2           1                3335.055664             0   
               3           1                   0.190723             0   
               4           1                   0.170940             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 0           1               -2.831221e-07   
               1           1                1.072884e-06   
               2           1                2.965331e-06   
               3           1                3.866851e-06   
               4           1                2.793968e-06   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 0           1                        0.012000   
               1           1                        0.011349   
               2           1                        0.008516   
               3           1                        0.015124   
               4           1                        1.743614   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 0           1                  1.126861         0.100603   
               1           1                  0.955973         0.100824   
               2           1              32599.023438         0.103627   
               3           1                  0.958584         0.099558   
               4           1                  0.860577         0.102369   

                                   OFP1_NFC2_total  MF_NFE_total  ...  \
SeriesNumber   EventNumber DetNum                                 ...   
25220213171932 0           1             -0.000551     -0.110674  ...   
               1           1              0.000328    283.672760  ...   
               2           1              6.105113     -0.526939  ...   
               3           1             -0.000703      0.235978  ...   
               4           1             -0.004189     -0.103168  ...   

                                   PulseWidth50_NFE_out  RiseTime2_NFC1_in  \
SeriesNumber   EventNumber DetNum                                            
25220213171932 0           1                          1                  1   
               1           1                         85                  1   
               2           1                          1                  0   
               3           1                          3                  3   
               4           1                          3                  1   

                                   IntegralTail_NFH_out  \
SeriesNumber   EventNumber DetNum                         
25220213171932 0           1                  -0.186589   
               1           1                  -0.664865   
               2           1                  -0.243079   
               3           1                  -0.047720   
               4           1                   0.119968   

                                   OFL_chi2timeFiltered_NFH_total  trig_ch  \
SeriesNumber   

In [5]:
data2.head()

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 100000      1       -4.148723e-07                    0.041316   
               100001      1        1.264224e-06                    1.896442   
               100002      1        6.785716e-07                    1.877393   
               100003      1        1.904996e-06                   -0.023704   
               100004      1        1.338213e-06                   -0.023486   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 100000      1                5970.476562             0   
               100001      1                   0.198083             0   
               100002      1                   0.206311             0   
               100003      1                   0.198820             0   
               100004      1                   0.194078             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 100000      1                    0.000003   
               100001      1                    0.000005   
               100002      1                   -0.000009   
               100003      1                    0.000003   
               100004      1                   -0.000006   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 100000      1                        0.006172   
               100001      1                        0.006973   
               100002      1                        0.007121   
               100003      1                        0.004459   
               100004      1                        0.006818   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 100000      1              44478.390625         0.097207   
               100001      1                  0.996291         0.096886   
               100002      1                  1.039752         0.097044   
               100003      1                  1.000425         0.098718   
               100004      1                  0.975441         0.098908   

                                   OFP1_NFC2_total  MF_NFE_total  ...  \
SeriesNumber   EventNumber DetNum                                 ...   
25220213171932 100000      1              4.113294      0.007574  ...   
               100001      1              0.001874     -0.052571  ...   
               100002      1             -0.003136      0.076841  ...   
               100003      1             -0.006587    216.476883  ...   
               100004      1              0.004911    471.424286  ...   

                                   PulseWidth50_NFE_out  RiseTime2_NFC1_in  \
SeriesNumber   EventNumber DetNum                                            
25220213171932 100000      1                          2                  0   
               100001      1                          2                  1   
               100002      1                          1                  0   
               100003      1                         63                  1   
               100004      1                        169                  1   

                                   IntegralTail_NFH_out  \
SeriesNumber   EventNumber DetNum                         
25220213171932 100000      1                   0.895641   
               100001      1                  33.681458   
               100002      1                  29.448662   
               100003      1                   0.059609   
               100004      1                  -0.292466   

                                   OFL_chi2timeFiltered_NFH_total  trig_ch  \
SeriesNumber   

In [6]:
data3.head()

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 200000      1       -1.256865e-07                   -0.024658   
               200001      1       -1.091389e-06                    0.014213   
               200002      1        1.346082e-07                   -0.023448   
               200003      1       -1.527056e-06                    1.418014   
               200004      1       -5.057776e-07                   -0.022999   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 200000      1               33481.242188             0   
               200001      1                1367.802856             0   
               200002      1                   0.203052             0   
               200003      1                   0.177843             0   
               200004      1                   0.227409             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 200000      1                6.198883e-06   
               200001      1               -1.716614e-05   
               200002      1               -4.693866e-07   
               200003      1                4.649162e-06   
               200004      1               -2.078712e-06   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 200000      1                        0.005344   
               200001      1                        0.005484   
               200002      1                        0.003419   
               200003      1                        0.004612   
               200004      1                        0.004263   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 200000      1             175754.625000         0.097022   
               200001      1              21440.787109         0.097575   
               200002      1                  1.021165         0.097141   
               200003      1                  0.896007         0.100108   
               200004      1                  1.145256         0.099239   

                                   OFP1_NFC2_total  MF_NFE_total  ...  \
SeriesNumber   EventNumber DetNum                                 ...   
25220213171932 200000      1              6.048519     -0.050356  ...   
               200001      1              5.088087      0.091596  ...   
               200002      1              0.011123     -0.235399  ...   
               200003      1             -0.012306     -0.078428  ...   
               200004      1             -0.000571    377.419952  ...   

                                   PulseWidth50_NFE_out  RiseTime2_NFC1_in  \
SeriesNumber   EventNumber DetNum                                            
25220213171932 200000      1                          3                  0   
               200001      1                          4                  1   
               200002      1                          2                  1   
               200003      1                          2                  1   
               200004      1                        125                  0   

                                   IntegralTail_NFH_out  \
SeriesNumber   EventNumber DetNum                         
25220213171932 200000      1                  -0.717270   
               200001      1                 -67.491516   
               200002      1                  -0.090750   
               200003      1                  19.384537   
               200004      1                  -0.554377   

                                   OFL_chi2timeFiltered_NFH_total  trig_ch  \
SeriesNumber   

In [7]:
!pip3 install memory_profiler
%load_ext memory_profiler

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/cvmfs/sft.cern.ch/lcg/views/LCG_98py3cu10/x86_64-centos7-gcc8-opt/bin/python -m pip install --upgrade pip' command.


## Vertical join

In vertical join, we stack dataframes one on top of the other, concatenating also the index.

In [8]:
%time vertical_join_df = pd.concat([data1, data2, data3], ignore_index = False, sort = False)

CPU times: user 290 ms, sys: 56 ms, total: 346 ms
Wall time: 345 ms


In [9]:
%memit pd.concat([data1, data2, data3], ignore_index = False, sort = False)

peak memory: 2029.45 MiB, increment: 495.78 MiB


In [10]:
vertical_join_df

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 0           1        9.468837e-08                   -0.024721   
               1           1       -2.323795e-06                   -0.023009   
               2           1        1.909242e-06                   -0.028883   
               3           1       -5.892106e-07                   -0.022784   
               4           1       -1.040180e-03                   -0.026172   
...                                          ...                         ...   
               299995      1       -8.445820e-07                   -0.022291   
               299996      1       -1.376572e-03                   -0.024523   
               299997      1        4.636416e-08                   -0.024157   
               299998      1       -1.376572e-03                   -0.024523   
               299999      1        1.361805e-06                   -0.023516   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 0           1                   0.224138             0   
               1           1                   0.190206             0   
               2           1                3335.055664             0   
               3           1                   0.190723             0   
               4           1                   0.170940             0   
...                                                 ...           ...   
               299995      1                   0.203781             0   
               299996      1                   0.171053             0   
               299997      1                   0.218077             0   
               299998      1                   0.171053             0   
               299999      1                1659.454102             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 0           1               -2.831221e-07   
               1           1                1.072884e-06   
               2           1                2.965331e-06   
               3           1                3.866851e-06   
               4           1                2.793968e-06   
...                                                  ...   
               299995      1                4.671514e-06   
               299996      1               -2.808869e-06   
               299997      1               -1.855195e-06   
               299998      1               -2.808869e-06   
               299999      1               -2.682209e-07   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 0           1                        0.012000   
               1           1                        0.011349   
               2           1                        0.008516   
               3           1                        0.015124   
               4           1                        1.743614   
...                                                      ...   
               299995      1                        0.011025   
               299996      1                        2.990817   
               299997      1                        0.009095   
               299998      1                        2.990817   
               299999      1                        0.011594   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 0           1                  1.126861         0.100603   
               1           1                  0.955973         0.100824   
               2           1              32599.023438         0.103627   
               3           1                  0.958584 

## Vertical join with different columns


In [3]:
data2_prefix = data2.add_prefix("data2_")

In [5]:
%time _ = pd.concat([data1, data2_prefix], ignore_index = False, sort = False)

CPU times: user 229 ms, sys: 103 ms, total: 332 ms
Wall time: 330 ms


## Vertical join + sort

I tried `sort=True` in `pd.concat` but doesn't work as I expected, so I use `sort_index` separately

In [11]:
data1_unordered = data1.sample(frac=1)
data2_unordered = data2.sample(frac=1)
data3_unordered = data3.sample(frac=1)

In [12]:
%time vertical_join_df_sort = pd.concat([data1_unordered, data2_unordered, data3_unordered], ignore_index = False, sort=False)

CPU times: user 279 ms, sys: 128 ms, total: 407 ms
Wall time: 407 ms


In [13]:
%time vertical_join_df_sort = vertical_join_df_sort.sort_index()

CPU times: user 572 ms, sys: 127 ms, total: 699 ms
Wall time: 701 ms


In [14]:
vertical_join_df_sort

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 0           1        9.468837e-08                   -0.024721   
               1           1       -2.323795e-06                   -0.023009   
               2           1        1.909242e-06                   -0.028883   
               3           1       -5.892106e-07                   -0.022784   
               4           1       -1.040180e-03                   -0.026172   
...                                          ...                         ...   
               299995      1       -8.445820e-07                   -0.022291   
               299996      1       -1.376572e-03                   -0.024523   
               299997      1        4.636416e-08                   -0.024157   
               299998      1       -1.376572e-03                   -0.024523   
               299999      1        1.361805e-06                   -0.023516   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 0           1                   0.224138             0   
               1           1                   0.190206             0   
               2           1                3335.055664             0   
               3           1                   0.190723             0   
               4           1                   0.170940             0   
...                                                 ...           ...   
               299995      1                   0.203781             0   
               299996      1                   0.171053             0   
               299997      1                   0.218077             0   
               299998      1                   0.171053             0   
               299999      1                1659.454102             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 0           1               -2.831221e-07   
               1           1                1.072884e-06   
               2           1                2.965331e-06   
               3           1                3.866851e-06   
               4           1                2.793968e-06   
...                                                  ...   
               299995      1                4.671514e-06   
               299996      1               -2.808869e-06   
               299997      1               -1.855195e-06   
               299998      1               -2.808869e-06   
               299999      1               -2.682209e-07   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 0           1                        0.012000   
               1           1                        0.011349   
               2           1                        0.008516   
               3           1                        0.015124   
               4           1                        1.743614   
...                                                      ...   
               299995      1                        0.011025   
               299996      1                        2.990817   
               299997      1                        0.009095   
               299998      1                        2.990817   
               299999      1                        0.011594   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 0           1                  1.126861         0.100603   
               1           1                  0.955973         0.100824   
               2           1              32599.023438         0.103627   
               3           1                  0.958584 

## Horizontal join

In horizontal join instead we have 2 dataframes that share the same index, at least part of it, and we want to align the columns.
Here I am sampling 80% of the rows and then sorting them, so that we do not have identical indices.

In [15]:
## If index is the same, join is crazy fast

%time _ = data1.join(data1.copy(), rsuffix="copy")

CPU times: user 126 ms, sys: 6 µs, total: 126 ms
Wall time: 125 ms


In [16]:
data1_80perc = data1.sample(frac=.8).sort_index()

In [17]:
%memit data1_join_data1 = data1.join(data1_80perc, rsuffix="copy")

peak memory: 3956.15 MiB, increment: 404.98 MiB


In [18]:
data1_join_data1.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 100000 entries, (25220213171932, 0, 1) to (25220213171932, 99999, 1)
Columns: 1050 entries, Slope_NFC1_in to IntegralTail_NFC2_incopy
dtypes: float32(810), float64(138), int16(5), int32(97)
memory usage: 457.6 MB


In [19]:
%time data1_join_data1 = data1.join(data1_80perc, rsuffix="copy")

CPU times: user 15.3 s, sys: 1.98 s, total: 17.3 s
Wall time: 17.3 s


In [20]:
data1_join_data1

Slope_NFC1_in  PlateauAmplitude_NFH_total  \
SeriesNumber   EventNumber DetNum                                              
25220213171932 0           1        9.468837e-08                   -0.024721   
               1           1       -2.323795e-06                   -0.023009   
               2           1        1.909242e-06                   -0.028883   
               3           1       -5.892106e-07                   -0.022784   
               4           1       -1.040180e-03                   -0.026172   
...                                          ...                         ...   
               99995       1        1.495558e-06                    1.917148   
               99996       1       -9.260073e-07                   -0.021705   
               99997       1        1.242497e-06                   -0.025264   
               99998       1       -2.518434e-07                   -0.023452   
               99999       1        1.236190e-06                   -0.023650   

                                   OFL_chi2b_NFC2_total  trig_pileups  \
SeriesNumber   EventNumber DetNum                                       
25220213171932 0           1                   0.224138             0   
               1           1                   0.190206             0   
               2           1                3335.055664             0   
               3           1                   0.190723             0   
               4           1                   0.170940             0   
...                                                 ...           ...   
               99995       1                   0.197972             0   
               99996       1                   0.187843             0   
               99997       1                2577.356201             0   
               99998       1                1134.245239             0   
               99999       1                   0.193317             0   

                                   IntegralHead_NFC2_out  \
SeriesNumber   EventNumber DetNum                          
25220213171932 0           1               -2.831221e-07   
               1           1                1.072884e-06   
               2           1                2.965331e-06   
               3           1                3.866851e-06   
               4           1                2.793968e-06   
...                                                  ...   
               99995       1               -1.303852e-06   
               99996       1               -3.933907e-06   
               99997       1               -1.184642e-06   
               99998       1               -2.399087e-06   
               99999       1                5.245209e-06   

                                   PlateauAmplitude_NFC1_out  \
SeriesNumber   EventNumber DetNum                              
25220213171932 0           1                        0.012000   
               1           1                        0.011349   
               2           1                        0.008516   
               3           1                        0.015124   
               4           1                        1.743614   
...                                                      ...   
               99995       1                        0.007164   
               99996       1                        0.005944   
               99997       1                        0.006445   
               99998       1                        0.007186   
               99999       1                        0.006020   

                                   OF0_chi2_NFC2_total  MeanBase_NFE_in  \
SeriesNumber   EventNumber DetNum                                         
25220213171932 0           1                  1.126861         0.100603   
               1           1                  0.955973         0.100824   
               2           1              32599.023438         0.103627   
               3           1                  0.958584 

## Horizontal join with no multiindex

Check if multi-index is making joins slower, it doesn't seem so

In [21]:
data1_nomultiindex = data1.reset_index(drop=True) 
data1_80perc_nomultiindex = data1_80perc.reset_index(drop=True) # I drop the index so we should have the same amount of memory

In [22]:
%time data1_join_data1 = data1.join(data1_80perc, rsuffix="copy")

CPU times: user 15.3 s, sys: 2.01 s, total: 17.3 s
Wall time: 17.3 s


In [23]:
%memit data1_join_data1_no_multiindex = data1_nomultiindex.join(data1_80perc_nomultiindex, rsuffix="copy")

peak memory: 5183.34 MiB, increment: 529.75 MiB


The memory profiler result I think are not very accurate, not sure why this operation should take 530 MB while the same with multi index 400MB, less than the size of the new dataframe!

In [24]:
data1_join_data1_no_multiindex.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 1050 entries, Slope_NFC1_in to IntegralTail_NFC2_incopy
dtypes: float32(810), float64(138), int16(5), int32(97)
memory usage: 452.2 MB


In [25]:
%time data1_join_data1_no_multiindex = data1_nomultiindex.join(data1_80perc_nomultiindex, rsuffix="copy")

CPU times: user 3.76 s, sys: 2.02 s, total: 5.78 s
Wall time: 5.81 s
